README.md
README in progress, for now please refer to 
- info@childgrowthmonitor.org
- [GitHub main project](https://github.com/Welthungerhilfe/ChildGrowthMonitor/)
- [Child Growth Monitor Website](https://childgrowthmonitor.org)

# Child Growth Monitor Machine Learning

## Introduction
This project uses machine learning to identify malnutrition from 3D scans of children under 5 years of age. This [one-minute video](https://www.youtube.com/watch?v=f2doV43jdwg) explains.

## Getting started

### Requirements
Training the models realistically requires using GPU computation. Project members are currently using a variety of cloud instances (GCS, AWS, Azure) and local machines for training. A separate backend project is currently developing the DevOps infrastructure to simplify this.

You will need:
* Python 3
* TensorFlow GPU
* Keras

### Installation
These steps provide an example installation on a local Ubuntu workstation from scratch:
* Install Ubuntu Desktop 18.04.1 LTS	
* Install NVIDIA drivers  
*Please note that after rebooting, the secure boot process will prompt you to authorize the driver to use the hardware via a MOK Management screen.*
```
sudo add-apt-repository ppa:graphics-drivers
sudo apt-get update
sudo apt-get install nvidia-390
sudo reboot now
```
* Install [Anaconda with Python 3.6](https://www.anaconda.com/download)
```conda update conda
conda update anaconda
conda update python
conda update --all
conda create --name cgm
source activate cgm
conda install tensorflow-gpu
conda install ipykernel
conda install keras
conda install vtk progressbar2 glob2 numbs pandas
pip install --upgrade pip
pip install git+https://github.com/daavoo/pyntcloud
```

### Dataset access
Data access is provided on as-needed basis following signature of the Welthungerhilfe Data Privacy & Commitment to
Maintain Data Secrecy Agreement. If you need data access (e.g. to train your machine learning models), 
please contact [Markus Matiaschek](mailto:mmatiaschek@gmail.com) for details.

### Preparing training data
* Run `python create_pickle.py`
* Run the notebook `create_dataset.ipynb`

### Training the models
* Run the notebook `train_on_dataset.ipynb`

*This currently takes around 6 hours to run on a local NVIDIA GTX 1080 Ti.*

### Evaluating the results
TODO

## Contributing

Please see [CONTRIBUTING.md](CONTRIBUTING.md) for details.

## Versioning

Our [releases](https://github.com/Welthungerhilfe/cgm-ml/releases) use [semantic versioning](http://semver.org). You can find a chronologically ordered list of notable changes in [CHANGELOG.md](CHANGELOG.md).

## License

This project is licensed under the GNU General Public License v3.0. See [LICENSE](LICENSE) for details and refer to [NOTICE](NOTICE) for additional licensing notes and use of third-party components.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

## 0.1.0

### Added
- Dataset creation
- Point cloud model
- Voxel grid model
- Model evaluation

### Changed
- Dataset creation now uses multiple processors (c83819f)

### Bugs
- n/a


tests/test_generator.py
import unittest
from datagenerator import DataGenerator, get_dataset_path, create_datagenerator_from_parameters


class TestGenerator(unittest.TestCase):

    def test_pointcloud_generation(self):
        dataset_path = get_dataset_path()

        dataset_parameters_pointclouds = {}
        dataset_parameters_pointclouds["input_type"] = "pointcloud"
        dataset_parameters_pointclouds["output_targets"] = ["height", "weight"]
        dataset_parameters_pointclouds["random_seed"] = 666
        dataset_parameters_pointclouds["pointcloud_target_size"] = 30000
        dataset_parameters_pointclouds["pointcloud_random_rotation"] = True
        dataset_parameters_pointclouds["dataset_size_train"] = 1000
        dataset_parameters_pointclouds["dataset_size_test"] = 20

        data_generator = create_datagenerator_from_parameters(dataset_path, dataset_parameters_pointclouds)
        dataset = next(data_generator.generate(size=1, yield_file_paths=True, verbose=True))


    def test_voxelgrid_generation(self):
        dataset_path = get_dataset_path()

        dataset_parameters_voxelgrids = {}
        dataset_parameters_voxelgrids["input_type"] = "voxelgrid"
        dataset_parameters_voxelgrids["output_targets"] = ["height", "weight"]
        dataset_parameters_voxelgrids["random_seed"] = 666
        dataset_parameters_voxelgrids["voxelgrid_target_shape"] = (32, 32, 32)
        dataset_parameters_voxelgrids["voxel_size_meters"] = 0.1
        dataset_parameters_voxelgrids["voxelgrid_random_rotation"] = True
        dataset_parameters_voxelgrids["dataset_size_train"] = 6000
        dataset_parameters_voxelgrids["dataset_size_test"] = 1000

        data_generator = create_datagenerator_from_parameters(dataset_path, dataset_parameters_voxelgrids)
        dataset = next(data_generator.generate(size=1, yield_file_paths=True, verbose=True))


if __name__ == '__main__':
    unittest.main()


seldon-deploy/test_post.py
import requests
import json
import numpy as np

addr = 'http://localhost:5000'
test_url = addr + "/cgm/regressor"

# Call server.
points = np.random.random(30000 * 4).astype("float32")
response = requests.post(test_url, data=points.tostring())#, headers=headers)
print(json.loads(response.text))


utils.py
import matplotlib.pyplot as plt
from mpl_toolkits.mplot3d import Axes3D
import numpy as np
import itertools
import datetime
import vtk
from mpl_toolkits.mplot3d import Axes3D
import glob
import os


def load_vtk(vtk_path):
    """
    Loads a VTK-file.
    """

    reader = vtk.vtkDataSetReader()
    reader.SetFileName(vtk_path)
    reader.ReadAllVectorsOn()
    reader.ReadAllScalarsOn()
    reader.Update()
    data = reader.GetOutput()

    points = np.zeros((data.GetNumberOfPoints(), 3))

    for i in range(data.GetNumberOfPoints()):
        points[i] = data.GetPoint(i)

    return points


def render_pointcloud(points, title=None):
    """
    Renders a point-cloud.
    """

    fig = plt.figure(figsize=(10, 10))
    ax = fig.add_subplot(111, projection='3d')

    ax.scatter(points[:,0], points[:,1], points[:,2], s=0.5, cmap="gray", alpha=0.5)

    ax.set_xlabel("x")
    ax.set_ylabel("y")
    ax.set_zlabel("z")

    if title != None:
        plt.title(title)

    plt.show()
    plt.close()


def ensure_voxelgrid_shape(voxelgrid, voxelgrid_target_shape):
    voxelgrid = pad_voxelgrid(voxelgrid, voxelgrid_target_shape)
    voxelgrid = crop_voxelgrid(voxelgrid, voxelgrid_target_shape)
    return voxelgrid


def pad_voxelgrid(voxelgrid, voxelgrid_target_shape):

    pad_before = [0.0] * 3
    pad_after = [0.0] * 3
    for i in range(3):
        pad_before[i] = (voxelgrid_target_shape[i] - voxelgrid.shape[i]) // 2
        pad_before[i] = max(0, pad_before[i])
        pad_after[i] = voxelgrid_target_shape[i] - pad_before[i] - voxelgrid.shape[i]
        pad_after[i] = max(0, pad_after[i])
    voxelgrid = np.pad(
        voxelgrid,
        [(pad_before[0], pad_after[0]), (pad_before[1], pad_after[1]), (pad_before[2], pad_after[2])],
        'constant', constant_values=[(0, 0), (0, 0), (0, 0)]
    )

    return voxelgrid


def crop_voxelgrid(voxelgrid, voxelgrid_target_shape):

    while voxelgrid.shape[0] > voxelgrid_target_shape[0]:
        voxels_start = np.count_nonzero(voxelgrid[0,:,:] != 0.0)
        voxels_end = np.count_nonzero(voxelgrid[-1,:,:] != 0.0)
        if voxels_start > voxels_end:
            voxelgrid = voxelgrid[:-1,:,:]
        else:
            voxelgrid = voxelgrid[1:,:,:]

    while voxelgrid.shape[1] > voxelgrid_target_shape[1]:
        voxels_start = np.count_nonzero(voxelgrid[:,0,:] != 0.0)
        voxels_end = np.count_nonzero(voxelgrid[:,-1,:] != 0.0)
        if voxels_start > voxels_end:
            voxelgrid = voxelgrid[:,:-1,:]
        else:
            voxelgrid = voxelgrid[:,1:,:]

    while voxelgrid.shape[2] > voxelgrid_target_shape[2]:
        voxels_start = np.count_nonzero(voxelgrid[:,:,0] != 0.0)
        voxels_end = np.count_nonzero(voxelgrid[:,:,-1] != 0.0)
        if voxels_start > voxels_end:
            voxelgrid = voxelgrid[:,:,:-1]
        else:
            voxelgrid = voxelgrid[:,:,1:]

    return voxelgrid


def center_crop_voxelgrid(voxelgrid, voxelgrid_target_shape):

    # Center crop.
    crop_start = [0.0] * 3
    crop_end = [0.0] * 3
    for i in range(3):
        crop_start[i] = (voxelgrid.shape[i] - voxelgrid_target_shape[i]) // 2
        crop_start[i] = max(0, crop_start[i])
        crop_end[i] = target_shape[i] + crop_start[i]
    voxelgrid = voxelgrid[crop_start[0]:crop_end[0], crop_start[1]:crop_end[1], crop_start[2]:crop_end[2]]

    return voxelgrid


def render_voxelgrid(voxelgrid, title=None):
    """
    Renders a voxel-grid.
    """

    figsize = (5, 5)
    fig = plt.figure(figsize=figsize)
    ax = fig.gca(projection='3d')
    transformed_voxelgrid = np.flip(np.flip(voxelgrid, axis=2), axis=0)

    facecolors = np.zeros(transformed_voxelgrid.shape + (3,))
    for x, y, z in itertools.product(range(transformed_voxelgrid.shape[0]), range(transformed_voxelgrid.shape[1]), range(transformed_voxelgrid.shape[2])):
        color = (1.0 - y / 32)
        facecolors[x, y, z, 0] = color
        facecolors[x, y, z, 1] = color
        facecolors[x, y, z, 2] = color

    ax.voxels(transformed_voxelgrid, facecolors=facecolors, edgecolor="k")

    if title != None:
        plt.title(title)

    plt.show()
    plt.close()


def get_datetime_string():
    """
    Returns a datetime string.
    """

    return datetime.datetime.now().strftime("%Y%m%d-%H%M")


def get_latest_preprocessed_dataset(path=".", filter=""):
    """
    Retrieves the path of the latest preprocessed dataset. Takes into account a filter.
    """
    glob_search_path = os.path.join(path, "*.p")
    paths = [x for x in glob.glob(glob_search_path) if filter in x]
    if len(paths) == 0:
        raise Exception("No datasets found for filter", filter)
    return sorted(paths)[-1]


def get_latest_model(path=".", filter=""):
    """
    Retrieves the path of the latest preprocessed dataset. Takes into account a filter.
    """
    glob_search_path = os.path.join(path, "*.h5")
    paths = [x for x in glob.glob(glob_search_path) if filter in x]
    if len(paths) == 0:
        raise Exception("No models found for filter", filter)
    return sorted(paths)[-1]


modelutils.py
from keras import models, layers
import numpy as np
import tensorflow as tf

def create_dense_model(input_shape, output_size):

    model = models.Sequential(name="baseline-dense")
    model.add(layers.Flatten(input_shape=input_shape))
    model.add(layers.Dense(128, activation="relu"))
    model.add(layers.Dense(64, activation="relu"))
    model.add(layers.Dense(output_size))

    return model


def create_voxnet_model_small(input_shape, output_size):
    """ See: http://dimatura.net/publications/3dcnn_lz_maturana_scherer_icra15.pdf """

    #Trainable params: 301,378
    model = models.Sequential(name="C7-F32-P2-C5-F64-P2-D512")
    model.add(layers.Reshape(target_shape=input_shape + (1,), input_shape=input_shape))
    model.add(layers.Conv3D(32, (7, 7, 7), activation="relu"))
    model.add(layers.MaxPooling3D((4, 4, 4)))
    model.add(layers.Conv3D(64, (5, 5, 5), activation="relu"))
    model.add(layers.MaxPooling3D((2, 2, 2)))
    model.add(layers.Flatten())
    model.add(layers.Dense(512, activation="relu"))
    model.add(layers.Dense(output_size))

    return model


def create_voxnet_model_big(input_shape, output_size):
    """ See: http://dimatura.net/publications/3dcnn_lz_maturana_scherer_icra15.pdf """

    # Trainable params: 7,101,442
    model = models.Sequential(name="C7-F64-P4-D512")
    model.add(layers.Reshape(target_shape=input_shape + (1,), input_shape=input_shape))
    model.add(layers.Conv3D(64, (7, 7, 7), activation="relu"))
    model.add(layers.MaxPooling3D((4, 4, 4)))
    model.add(layers.Flatten())
    model.add(layers.Dense(512, activation="relu"))
    model.add(layers.Dense(output_size))

    return model


def create_voxnet_model_homepage(input_shape, output_size):
    """ See: http://dimatura.net/publications/3dcnn_lz_maturana_scherer_icra15.pdf """

    # Trainable params: 916,834
    model = models.Sequential(name="VoxNetHomepage")
    model.add(layers.Reshape(target_shape=input_shape + (1,), input_shape=input_shape))
    model.add(layers.Conv3D(32, (5, 5, 5), strides=(2, 2, 2), activation="relu"))
    model.add(layers.Conv3D(32, (3, 3, 3), strides=(1, 1, 1), activation="relu"))
    model.add(layers.MaxPooling3D((2, 2, 2)))
    model.add(layers.Flatten())
    model.add(layers.Dense(128, activation="relu"))
    model.add(layers.Dense(output_size))

    return model


def create_point_net(input_shape, output_size):
    """ See https://github.com/garyloveavocado/pointnet-keras/blob/master/train_cls.py """

    num_points = input_shape[0]

    input_points = layers.Input(shape=input_shape)
    x = layers.Convolution1D(64, 1, activation='relu',
                      input_shape=input_shape)(input_points)
    x = layers.BatchNormalization()(x)
    x = layers.Convolution1D(128, 1, activation='relu')(x)
    x = layers.BatchNormalization()(x)
    x = layers.Convolution1D(1024, 1, activation='relu')(x)
    x = layers.BatchNormalization()(x)
    x = layers.MaxPooling1D(pool_size=num_points)(x)
    x = layers.Dense(512, activation='relu')(x)
    x = layers.BatchNormalization()(x)
    x = layers.Dense(256, activation='relu')(x)
    x = layers.BatchNormalization()(x)
    x = layers.Dense(9, weights=[np.zeros([256, 9]), np.array([1, 0, 0, 0, 1, 0, 0, 0, 1]).astype(np.float32)])(x)
    input_T = layers.Reshape((3, 3))(x)

    # forward net
    g = layers.Lambda(mat_mul, arguments={'B': input_T})(input_points)
    g = layers.Convolution1D(64, 1, input_shape=input_shape, activation='relu')(g)
    g = layers.BatchNormalization()(g)
    g = layers.Convolution1D(64, 1, input_shape=input_shape, activation='relu')(g)
    g = layers.BatchNormalization()(g)

    # feature transform net
    f = layers.Convolution1D(64, 1, activation='relu')(g)
    f = layers.BatchNormalization()(f)
    f = layers.Convolution1D(128, 1, activation='relu')(f)
    f = layers.BatchNormalization()(f)
    f = layers.Convolution1D(1024, 1, activation='relu')(f)
    f = layers.BatchNormalization()(f)
    f = layers.MaxPooling1D(pool_size=num_points)(f)
    f = layers.Dense(512, activation='relu')(f)
    f = layers.BatchNormalization()(f)
    f = layers.Dense(256, activation='relu')(f)
    f = layers.BatchNormalization()(f)
    f = layers.Dense(64 * 64, weights=[np.zeros([256, 64 * 64]), np.eye(64).flatten().astype(np.float32)])(f)
    feature_T = layers.Reshape((64, 64))(f)

    # forward net
    g = layers.Lambda(mat_mul, arguments={'B': feature_T})(g)
    g = layers.Convolution1D(64, 1, activation='relu')(g)
    g = layers.BatchNormalization()(g)
    g = layers.Convolution1D(128, 1, activation='relu')(g)
    g = layers.BatchNormalization()(g)
    g = layers.Convolution1D(1024, 1, activation='relu')(g)
    g = layers.BatchNormalization()(g)

    # global_feature
    global_feature = layers.MaxPooling1D(pool_size=num_points)(g)

    # point_net_cls
    c = layers.Dense(512, activation='relu')(global_feature)
    c = layers.BatchNormalization()(c)
    c = layers.Dropout(rate=0.7)(c)
    c = layers.Dense(256, activation='relu')(c)
    c = layers.BatchNormalization()(c)
    c = layers.Dropout(rate=0.7)(c)
    #c = layers.Dense(k, activation='softmax')(c)
    c = layers.Dense(output_size, activation='linear')(c)
    prediction = layers.Flatten()(c)

    model = models.Model(inputs=input_points, outputs=prediction)
    return model

def mat_mul(A, B):
    return tf.matmul(A, B)


seldon-deploy/regressor.py
from __future__ import absolute_import

# For importing stuff.
import sys
sys.path.append("..")

import utils
from keras.models import load_model
import numpy as np
import pyntcloud
import pandas as pd
from flask import Flask, request, Response
import jsonpickle
import tensorflow as tf
from keras import backend as K


class CGMRegressor(object):

    def __init__(self):
        model_path = utils.get_latest_model("..", "voxnet")
        print(model_path)

        self.model = load_model(model_path)
        self.model.summary()
        self.graph = tf.get_default_graph()

        self.voxel_size_meters = 0.1


    def predict(self, X, features_names):

        # Reshape first.
        points = np.reshape(X, (-1, 4))

        # Turn it into a voxelgrid.
        dataframe = pd.DataFrame(points, columns=["x", "y", "z", "c"])
        point_cloud = pyntcloud.PyntCloud(dataframe)
        voxelgrid_id = point_cloud.add_structure("voxelgrid", size_x=self.voxel_size_meters, size_y=self.voxel_size_meters, size_z=self.voxel_size_meters)
        voxelgrid = point_cloud.structures[voxelgrid_id].get_feature_vector(mode="density")

        # Pad the voxelgrid.
        voxelgrid = utils.ensure_voxelgrid_shape(voxelgrid, (32, 32, 32))
        voxelgrid = np.expand_dims(voxelgrid, axis=0)

        with self.graph.as_default():
            return self.model.predict(voxelgrid)


# When using flask.




# Note: Just for testing. Hopefully seldon will do something similar.
if __name__ == "__main__":

    regressor = CGMRegressor()
    app = Flask(__name__)

    @app.route("/cgm/regressor", methods=["POST"])
    def flask_post():

        r = request
        points = np.fromstring(r.data, np.float32)
        print("Received points", points.shape)

        # Predict.
        prediction = regressor.predict(points, [])[0]

        # Build the response.
        response = {
            "message": "success",
            "height": str(prediction[0]),
            "weight": str(prediction[1])
        }
        response = jsonpickle.encode(response)

        # Respond.
        return Response(response=response, status=200, mimetype="application/json")

    app.run(host="0.0.0.0", port=5000)

    #pointcloud = np.random.random(30000 * 4)
    #prediction = regressor.predict(pointcloud, features_names=["height", "weight"])
    #print("Prediction:", prediction)


create_pickle.py
import os
from datagenerator import DataGenerator
import pickle
import random
import datetime

if os.path.exists("datasetpath.txt"):
    dataset_path = open("datasetpath.txt", "r").read().replace("\n", "")
else:
    dataset_path = "/home/jovyan/work/data/"

input_type="pointcloud"

# For creating voxelgrids.
if input_type == "voxelgrid":
    dataset_parameters = {}
    dataset_parameters["input_type"] = "voxelgrid"
    dataset_parameters["output_targets"] = ["height", "weight"]    
    dataset_parameters["random_seed"] = 666
    dataset_parameters["voxelgrid_target_shape"] = (32, 32, 32)
    dataset_parameters["voxel_size_meters"] = 0.1
    dataset_parameters["voxelgrid_random_rotation"] = True
    dataset_parameters["dataset_size_train"] = 6000
    dataset_parameters["dataset_size_test"] = 1000

# For creating pointclouds.
if input_type == "pointcloud":
    dataset_parameters = {}
    dataset_parameters["input_type"] = "pointcloud"
    dataset_parameters["output_targets"] = ["height", "weight"]    
    dataset_parameters["random_seed"] = 666
    dataset_parameters["pointcloud_target_size"] = 30000
    dataset_parameters["pointcloud_random_rotation"] = True
    dataset_parameters["dataset_size_train"] = 3000
    dataset_parameters["dataset_size_test"] = 500

print("Creating data-generator...")
data_generator = DataGenerator(
    dataset_path=dataset_path, 
    input_type=dataset_parameters["input_type"], 
    output_targets=dataset_parameters["output_targets"],
    voxelgrid_target_shape=dataset_parameters.get("voxelgrid_target_shape", None),
    voxel_size_meters=dataset_parameters.get("voxel_size_meters", None),
    voxelgrid_random_rotation=dataset_parameters.get("voxelgrid_random_rotation", None),
    pointcloud_target_size=dataset_parameters.get("pointcloud_target_size", None),
    pointcloud_random_rotation=dataset_parameters.get("pointcloud_random_rotation", None)
)
data_generator.print_statistics()

do_analysis = False
#do_analysis = True

if do_analysis == True:
    data_generator.analyze_files()
    data_generator.analyze_targets()
    data_generator.analyze_pointclouds()
    data_generator.analyze_voxelgrids()
    # how much data per measure?
else:
    print("Skipped analysis.")

# Do the split.
random.seed(dataset_parameters["random_seed"])
qrcodes_shuffle = data_generator.qrcodes[:]
random.shuffle(qrcodes_shuffle)
split_index = int(0.8 * len(qrcodes_shuffle))
qrcodes_train = sorted(qrcodes_shuffle[:split_index])
qrcodes_test = sorted(qrcodes_shuffle[split_index:])
del qrcodes_shuffle
print("")

print("QR-Codes for training:", " ".join(qrcodes_train))
print("")
print("QR-Codes for testing:", " ".join(qrcodes_test))
print("")

print("Generating training data...")
dataset_train = next(data_generator.generate(size=dataset_parameters["dataset_size_train"], verbose=False))

print("Generating testing data...")
dataset_test = next(data_generator.generate(size=dataset_parameters["dataset_size_test"], verbose=False))
    
print("Done.")

datetime_string = datetime.datetime.now().strftime("%Y%m%d-%H%M")
dataset_name = datetime_string + "-" + dataset_parameters["input_type"] + "-dataset.p"
pickle.dump((dataset_train, dataset_test, dataset_parameters), open(dataset_name, "wb"))
print("Saved " + dataset_name)